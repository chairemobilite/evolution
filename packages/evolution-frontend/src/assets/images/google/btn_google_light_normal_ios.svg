<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 22.0.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_1" xmlns:sketch="http://www.bohemiancoding.com/sketch/ns"
	 xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" viewBox="0 0 18 18.5"
	 style="enable-background:new 0 0 18 18.5;" xml:space="preserve">
<style type="text/css">
	.st0{fill:#4285F4;}
	.st1{fill:#34A853;}
	.st2{fill:#FBBC05;}
	.st3{fill:#EA4335;}
	.st4{fill:none;}
</style>
<title>btn_google_light_normal_ios</title>
<desc>Created with Sketch.</desc>
<g id="_x39_-PATCH" transform="translate(-608.000000, -160.000000)" sketch:type="MSArtboardGroup">
</g>
<g id="logo_googleg_48dp" transform="translate(15.000000, 15.000000)" sketch:type="MSLayerGroup">
	<path id="Shape" sketch:type="MSShapeGroup" class="st0" d="M2.7-5.6c0-0.6-0.1-1.3-0.2-1.8h-8.5v3.5h4.8c-0.2,1.1-0.8,2.1-1.8,2.7
		V1h2.9C1.8-0.5,2.7-2.8,2.7-5.6L2.7-5.6z"/>
	<path id="Shape_1_" sketch:type="MSShapeGroup" class="st1" d="M-5.9,3.2c2.4,0,4.5-0.8,6-2.2l-2.9-2.3c-0.8,0.5-1.8,0.9-3,0.9
		c-2.3,0-4.3-1.6-5-3.7h-3v2.3C-12.5,1.2-9.4,3.2-5.9,3.2L-5.9,3.2z"/>
	<path id="Shape_2_" sketch:type="MSShapeGroup" class="st2" d="M-10.9-4.1c-0.2-0.5-0.3-1.1-0.3-1.7s0.1-1.2,0.3-1.7v-2.3h-3
		c-0.6,1.2-1,2.6-1,4s0.3,2.8,1,4L-10.9-4.1L-10.9-4.1z"/>
	<path id="Shape_3_" sketch:type="MSShapeGroup" class="st3" d="M-5.9-11.2c1.3,0,2.5,0.5,3.4,1.3l2.6-2.6c-1.6-1.5-3.6-2.3-6-2.3
		c-3.5,0-6.6,2-8,5l3,2.3C-10.2-9.6-8.2-11.2-5.9-11.2L-5.9-11.2z"/>
	<path id="Shape_4_" sketch:type="MSShapeGroup" class="st4" d="M-14.9-14.8h18v18h-18V-14.8z"/>
</g>
<g id="handles_square" sketch:type="MSLayerGroup">
</g>
</svg>
