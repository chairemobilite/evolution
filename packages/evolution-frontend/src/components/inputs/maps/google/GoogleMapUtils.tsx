/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { FeatureGeocodedProperties } from '../InputMapTypes';
import { point as turfPoint } from '@turf/turf';

export const geojson = (
    latLng: google.maps.LatLng | null | undefined
): GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties> | undefined => {
    return latLng && latLng.lat() && latLng.lng() ? turfPoint([latLng.lng(), latLng.lat()], {}) : undefined;
};

export const toLatLng = (feature: GeoJSON.Feature<GeoJSON.Point>): google.maps.LatLng => {
    return new google.maps.LatLng(feature.geometry.coordinates[1], feature.geometry.coordinates[0]);
};
