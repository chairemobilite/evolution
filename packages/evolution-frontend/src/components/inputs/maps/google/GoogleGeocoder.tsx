/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { FeatureGeocodedProperties, PlaceGeocodedProperties } from '../InputMapTypes';
import { geojson } from './GoogleMapUtils';
import { point as turfPoint, distance as turfDistance } from '@turf/turf';

// In meters:
const geocodingSearchRadiusMinimum = 500;
const geocodingSearchRadiusMaximum = 8000;

export const geocodeSinglePoint = (
    addressQueryString: string,
    options: { bbox?: [number, number, number, number] }
): Promise<GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties> | undefined> => {
    return new Promise((resolve, reject) => {
        if (!google) {
            resolve(undefined);
        }
        const geocoder = new google.maps.Geocoder();
        const bounds = options.bbox
            ? new google.maps.LatLngBounds(
                { lat: options.bbox[0], lng: options.bbox[1] },
                { lat: options.bbox[2], lng: options.bbox[3] }
            )
            : undefined;
        geocoder.geocode({ address: addressQueryString, bounds }, (results, status) => {
            if (status === google.maps.GeocoderStatus.OK && results !== null) {
                const feature = geojson(results[0].geometry.location);
                if (feature) {
                    feature.properties.geocodingResultMetadata = {
                        formattedAddress: results[0].formatted_address,
                        precision: results[0].geometry.location_type,
                        types: results[0].types,
                        placeId: results[0].place_id
                    };
                    feature.properties.lastAction = 'geocoding';
                    feature.properties.geocodingQueryString = addressQueryString;
                }
                resolve(feature);
            } else {
                reject(`Geocoding failed for ${addressQueryString}`);
            }
        });
    });
};

export const geocodeMultiplePlaces = (
    geocodingQueryString: string,
    options: { bbox?: [number, number, number, number]; map?: google.maps.Map; language?: string }
): Promise<GeoJSON.Feature<GeoJSON.Point, PlaceGeocodedProperties>[] | undefined> => {
    return new Promise((resolve, reject) => {
        if (!options.map) {
            reject('Undefined map: Cannot use Google geocoder without Google maps');
            return;
        }
        if (!google || !options.bbox) {
            resolve(undefined);
            return;
        }
        const geocoder = new google.maps.places.PlacesService(options.map);
        const bounds = new google.maps.LatLngBounds(
            { lat: options.bbox[0], lng: options.bbox[1] },
            { lat: options.bbox[2], lng: options.bbox[3] }
        );

        const upperBound = turfPoint([options.bbox[0], options.bbox[1]]);
        const lowerBound = turfPoint([options.bbox[2], options.bbox[3]]);
        const viewportRadius = turfDistance(upperBound, lowerBound, { units: 'meters' }) / 2;
        const searchRadius = Math.min(
            Math.max(viewportRadius, geocodingSearchRadiusMinimum),
            geocodingSearchRadiusMaximum
        );

        geocoder.textSearch(
            {
                query: geocodingQueryString,
                location: bounds.getCenter(),
                radius: searchRadius,
                language: options.language
            },
            (results, status) => {
                if (status === google.maps.places.PlacesServiceStatus.OK && results !== null) {
                    const places = results
                        .map(({ geometry, ...googlePlace }) => {
                            const feature = geojson(geometry?.location) as GeoJSON.Feature<
                                GeoJSON.Point,
                                PlaceGeocodedProperties
                            >;
                            if (!feature) {
                                return undefined;
                            }
                            feature.id = googlePlace.place_id;
                            feature.properties.placeData = {
                                ...googlePlace
                            };
                            return feature;
                        })
                        .filter((feature) => feature !== undefined) as GeoJSON.Feature<
                        GeoJSON.Point,
                        PlaceGeocodedProperties
                    >[];

                    resolve(places);
                } else {
                    reject(`Geocoding failed for ${geocodingQueryString}`);
                }
            }
        );
    });
};
