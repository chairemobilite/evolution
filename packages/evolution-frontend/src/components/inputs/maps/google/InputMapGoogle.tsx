/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { GoogleMap, useJsApiLoader, Marker } from '@react-google-maps/api';
import GeoJSON from 'geojson';
import bowser from 'bowser';

import projectConfig from 'chaire-lib-common/lib/config/shared/project.config';
import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import { googleMapConfigNew as googleConfig } from '../../../../config/googleMaps.config';
import InputLoading from '../../InputLoading';
import { FeatureGeocodedProperties, MarkerData, InfoWindow } from '../InputMapTypes';
import { geojson, toLatLng } from './GoogleMapUtils';

export interface InputGoogleMapPointProps {
    defaultCenter: { lat: number; lon: number };
    maxGeocodingResultsBounds?: [{ lat: number; lng: number }, { lat: number; lng: number }];
    value?: GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties>;
    defaultZoom?: number;
    maxZoom?: number;
    height?: string; // the height of the map container in css units: example: 28rem or 550px
    markers: MarkerData[];
    onValueChange: (feature: GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties> | undefined) => void;
    onMapReady?: (bbox?: [number, number, number, number]) => void;
    onBoundsChanged?: (bbox?: [number, number, number, number]) => void;
    // Change this value when map bounds should be increased to fit markers
    shouldFitBounds?: number;
    infoWindow?: InfoWindow;
    setGeocodingOptions?: (options: { [key: string]: unknown }) => void;
}

const callWithBounds = (
    bounds?: google.maps.LatLngBounds,
    call?: (bbox?: [number, number, number, number]) => void
) => {
    if (!call) {
        return;
    }
    call(
        bounds
            ? [
                bounds.getSouthWest().lat(),
                bounds.getSouthWest().lng(),
                bounds.getNorthEast().lat(),
                bounds.getNorthEast().lng()
            ]
            : undefined
    );
};

const InputMapGoogle: React.FunctionComponent<InputGoogleMapPointProps> = (props: InputGoogleMapPointProps) => {
    const { isLoaded } = useJsApiLoader({
        region: projectConfig.region,
        language: projectConfig.defaultLocale,
        ...googleConfig
    });

    const [map, setMap] = React.useState<google.maps.Map | null>(null);
    const [center, setCenter] = React.useState<{ lat: number; lng: number } | google.maps.LatLng>({
        lat: props.defaultCenter.lat,
        lng: props.defaultCenter.lon
    });
    const [placesInfoWindow, setPlacesInfoWindow] = React.useState<google.maps.InfoWindow | undefined>(undefined);

    const changeValueAndPan = React.useCallback(
        (feature: GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties> | undefined) => {
            props.onValueChange(feature);
            if (feature) {
                setCenter({ lat: feature.geometry.coordinates[1], lng: feature.geometry.coordinates[0] });
                (map as google.maps.Map).panTo({
                    lat: feature.geometry.coordinates[1],
                    lng: feature.geometry.coordinates[0]
                });
            }
        },
        [map]
    );

    const onLoad = React.useCallback((map: google.maps.Map) => {
        setMap(map);
        if (props.setGeocodingOptions) props.setGeocodingOptions({ map });
        callWithBounds(map.getBounds(), props.onMapReady);
        setPlacesInfoWindow(
            new google.maps.InfoWindow({
                pixelOffset: new google.maps.Size(0, -40)
            })
        );
    }, []);

    const onUnmount = React.useCallback(() => {
        setMap(null);
        if (props.setGeocodingOptions) props.setGeocodingOptions({ map: undefined });
    }, []);

    const onPositionChange = React.useCallback(
        (e: google.maps.MapMouseEvent, triggerEvent: string) => {
            const geojsonValue = geojson(e.latLng);
            if (geojsonValue) {
                geojsonValue.properties.lastAction = triggerEvent;
                geojsonValue.properties.zoom = (map as google.maps.Map).getZoom();
                geojsonValue.properties.platform = bowser.getParser(window.navigator.userAgent).getPlatformType();
            }
            changeValueAndPan(geojsonValue);
        },
        [map]
    );

    const onZoomChange = () => {
        if (!map) return;
        const currentZoom = (map as google.maps.Map).getZoom();
        if (currentZoom && props.maxZoom && props.maxZoom < currentZoom) {
            (map as google.maps.Map).setZoom(props.maxZoom);
        }
    };

    const onBoundsChanged = () => {
        if (!map) return;
        callWithBounds((map as google.maps.Map).getBounds(), props.onBoundsChanged);
    };

    React.useEffect(() => {
        if (map) {
            (map as google.maps.Map).panTo({
                lat: props.defaultCenter.lat,
                lng: props.defaultCenter.lon
            });
        }
    }, [props.defaultCenter]);

    React.useEffect(() => {
        if (map && props.markers.length >= 1) {
            const maxMapBounds = new google.maps.LatLngBounds();
            if (props.maxGeocodingResultsBounds !== undefined) {
                maxMapBounds.extend(props.maxGeocodingResultsBounds[0]); // lower left
                maxMapBounds.extend(props.maxGeocodingResultsBounds[1]); // upper right
            }
            let atLeastOneMarkerInMaxBounds = false;
            const markerBounds = new google.maps.LatLngBounds();
            const boundsForMarkersInMaxBounds = new google.maps.LatLngBounds();
            props.markers.forEach((marker) => {
                const position = {
                    lat: marker.position.geometry.coordinates[1],
                    lng: marker.position.geometry.coordinates[0]
                };
                markerBounds.extend(position);

                if (!maxMapBounds.isEmpty() && maxMapBounds.contains(position)) {
                    atLeastOneMarkerInMaxBounds = true;
                    boundsForMarkersInMaxBounds.extend(position);
                }
            });

            const bounds = atLeastOneMarkerInMaxBounds ? boundsForMarkersInMaxBounds : markerBounds;

            // hack for single markers: see https://stackoverflow.com/questions/3334729/google-maps-v3-fitbounds-zoom-too-close-for-single-marker
            if (bounds.getNorthEast().equals(bounds.getSouthWest())) {
                const extendPoint1 = new google.maps.LatLng(
                    bounds.getNorthEast().lat() + 0.001,
                    bounds.getNorthEast().lng() + 0.001
                );
                const extendPoint2 = new google.maps.LatLng(
                    bounds.getNorthEast().lat() - 0.001,
                    bounds.getNorthEast().lng() - 0.001
                );
                bounds.extend(extendPoint1);
                bounds.extend(extendPoint2);
            }
            map.fitBounds(bounds);
        }
    }, [props.shouldFitBounds]);

    React.useEffect(() => {
        if (!placesInfoWindow) {
            return;
        }
        placesInfoWindow.close();
        if (props.infoWindow) {
            placesInfoWindow.setContent(props.infoWindow.content);
            placesInfoWindow.setPosition({
                lat: props.infoWindow.position.geometry.coordinates[1],
                lng: props.infoWindow.position.geometry.coordinates[0]
            });
            placesInfoWindow.open(map);
        }
    }, [props.infoWindow]);

    return isLoaded ? (
        <GoogleMap
            mapContainerStyle={{
                boxSizing: 'border-box' as const,
                position: 'relative' as const,
                width: '100%',
                height: props.height || '40rem',
                border: '1px solid rgba(0,0,0,0.2)'
            }}
            center={center}
            zoom={props.defaultZoom || 10}
            onLoad={onLoad}
            onUnmount={onUnmount}
            onClick={(e) => onPositionChange(e, 'mapClicked')}
            onZoomChanged={onZoomChange}
            onBoundsChanged={onBoundsChanged}
            clickableIcons={false}
        >
            {props.markers.map((markerData, index) => (
                <Marker
                    key={`marker${index}${markerData.position.geometry.coordinates[0]}_${markerData.position.geometry.coordinates[1]}`}
                    position={toLatLng(markerData.position)}
                    onDragEnd={markerData.draggable ? (e) => onPositionChange(e, 'markerDragged') : undefined}
                    draggable={markerData.draggable}
                    icon={{
                        url: markerData.icon.url,
                        size: new google.maps.Size(markerData.icon.size[0], markerData.icon.size[1]),
                        scaledSize: new google.maps.Size(markerData.icon.size[0], markerData.icon.size[1])
                    }}
                    onClick={markerData.onClick}
                />
            ))}
        </GoogleMap>
    ) : (
        <InputLoading />
    );
};

export default InputMapGoogle;
