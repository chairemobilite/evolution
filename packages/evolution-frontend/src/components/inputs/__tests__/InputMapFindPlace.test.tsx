/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { interviewAttributes } from './interviewData.test';
import InputMapFindPlace from '../InputMapFindPlace';
import { geocodeMultiplePlaces } from '../maps/google/GoogleGeocoder';

// Mock db queries
jest.mock('../maps/google/GoogleGeocoder', () => ({
    geocodeMultiplePlaces: jest.fn()
}));
const mockedGeocode = geocodeMultiplePlaces as jest.MockedFunction<typeof geocodeMultiplePlaces>;

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
}

const baseWidgetConfig = {
    type: 'question' as const,
    twoColumns: true,
    path: 'test.foo',
    containsHtml: true,
    label: {
        fr: `Texte en français`,
        en: `English text`
    },
    size: 'medium' as const,
    inputType: 'mapFindPlace' as const,
    defaultCenter: { lat: 45, lon: -73 },
    showPhoto: false
};

describe('Render InputMapPoint with various parameters', () => {

    test('Test with minimal parameters', () => {
        const wrapper = TestRenderer.create(
            <InputMapFindPlace
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={baseWidgetConfig}
                value={undefined}
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Test with all parameters', () => {
            
        const testWidgetConfig = Object.assign({
            geocodingQueryString: jest.fn(),
            refreshGeocodingLabel: {
                fr: `Rafraîchir la carte`,
                en: `Refresh map`
            },
            icon: {
                url: 'path/to/icon',
                size: [80, 80] as [number, number]
            },
            maxZoom: 18,
            defaultZoom: 15,
            coordinatesPrecision: 6,
            placesIcon: {
                url: 'path/to/icon',
                size: [85, 85] as [number, number]
            },
            maxGeocodingResultsBounds: function (interview, path) {
                return [{lat: 45.2229, lng: -74.3230}, {lat: 46.1181, lng: -72.9215}] as [{ lat: number; lng: number; }, { lat: number; lng: number; }];
            },
            updateDefaultValueWhenResponded: true
        }, baseWidgetConfig);
        const wrapper = TestRenderer.create(
            <InputMapFindPlace
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={testWidgetConfig}
                value={{ type: 'Feature' as const, properties: {}, geometry: { type: 'Point' as const, coordinates: [-73.1, 45.02]}}}
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
    
});

describe('Test geocoding requests', () => {
    const testId = 'test';
    const geocodingString = 'foo restaurant';
    const mockOnValueChange = jest.fn();

    const testWidgetConfig = Object.assign({
        geocodingQueryString: jest.fn().mockReturnValue(geocodingString),
        refreshGeocodingLabel: {
            fr: `Geocode`,
            en: `Geocode`
        },
        icon: {
            url: 'path/to/icon',
        },
        maxZoom: 18,
        defaultZoom: 15,
        placesIcon: {
            url: 'path/to/icon',
        },
        size: 'medium',
        updateDefaultValueWhenResponded: true
    }, baseWidgetConfig);

    const placeFeature1 = { 
        type: 'Feature' as const, 
        geometry: { type: 'Point' as const, coordinates: [-73.2, 45.1] },
        properties: { placeData: { place_id: 2, formatted_address: '123 test street', name: 'Foo extra good restaurant' } }
    };
    const placeFeature2 = {
        type: 'Feature' as const, 
        geometry: { type: 'Point' as const, coordinates: [-73.2, 45.1] },
        properties: { placeData: { place_id: 2, formatted_address: '123 foo street', types: ['street_address'] } }
    };

    beforeEach(() => {
        mockedGeocode.mockClear();
        mockOnValueChange.mockClear();
    });

    test('Geocode with multiple results', async () => {
    
        const findPlaceWidget = mount(<InputMapFindPlace
            id={testId}
            onValueChange={mockOnValueChange}
            widgetConfig={testWidgetConfig}
            value={{ type: 'Feature' as const, properties: {}, geometry: { type: 'Point' as const, coordinates: [-73.1, 45.02]}}}
            inputRef={React.createRef()}
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
            loadingState={0}
        />);

        // Find and click on the Geocode button, to return multiple values
        mockedGeocode.mockResolvedValueOnce([placeFeature1, placeFeature2]);
        const geocodeButton = findPlaceWidget.find({id: `${testId}_refresh`, type: 'button'});
        expect(geocodeButton.getDOMNode<HTMLButtonElement>().textContent).toBe('Geocode');
        geocodeButton.simulate('mousedown');
        geocodeButton.simulate('mouseup');
        // Let async functions terminate
        await new Promise(process.nextTick);
    
        // The select list should display 4 children (2 places and the 2 extra elements) and should not have a confirm button
        findPlaceWidget.update();
        expect(mockedGeocode).toHaveBeenCalledTimes(1);
        expect(mockedGeocode).toHaveBeenCalledWith(geocodingString, expect.anything());
        const selectionList = findPlaceWidget.find(`select`);
        const selectionDomElement = selectionList.getDOMNode<HTMLSelectElement>();
        expect(selectionDomElement.children.length).toEqual(4);
        expect(selectionDomElement.children[1].textContent).toEqual('Foo extra good restaurant (123 test street)');
        expect(selectionDomElement.children[2].textContent).toEqual('123 foo street');
    
        const confirmButton = findPlaceWidget.find({id: `${testId}_confirm`, type: 'button'});
        expect(confirmButton.length).toEqual(0);

        // Make sure the value has not been changed
        expect(mockOnValueChange).not.toHaveBeenCalled();
        
    });

    test('Geocode with single results, and confirm result', async () => {
        
        const findPlaceWidget = mount(<InputMapFindPlace
            id={testId}
            onValueChange={mockOnValueChange}
            widgetConfig={testWidgetConfig}
            value={{ type: 'Feature' as const, properties: {}, geometry: { type: 'Point' as const, coordinates: [-73.1, 45.02]}}}
            inputRef={React.createRef()}
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
            loadingState={0}
        />);
        
        // Find and click on the Geocode button, to return multiple values
        mockedGeocode.mockResolvedValueOnce([placeFeature1]);
        const geocodeButton = findPlaceWidget.find({id: `${testId}_refresh`, type: 'button'});
        expect(geocodeButton.getDOMNode<HTMLButtonElement>().textContent).toBe('Geocode');
        geocodeButton.simulate('mousedown');
        geocodeButton.simulate('mouseup');
        // Let async functions terminate
        await new Promise(process.nextTick);
    
        // The select list should display 3 children (1 place and the 2 extra elements) and the confirm button should be present
        findPlaceWidget.update();
        expect(mockedGeocode).toHaveBeenCalledTimes(1);
        expect(mockedGeocode).toHaveBeenCalledWith(geocodingString, expect.anything());
        const selectionList = findPlaceWidget.find(`select`);
        const selectionDomElement = selectionList.getDOMNode<HTMLSelectElement>();
        expect(selectionDomElement.children.length).toEqual(3);
        expect(selectionDomElement.children[1].textContent).toEqual('Foo extra good restaurant (123 test street)');
    
        // Click on the confirm button and make sure the onValueChanged function has been called
        const confirmButton = findPlaceWidget.find({id: `${testId}_confirm`, type: 'button'});
        expect(confirmButton.length).toEqual(1);
        confirmButton.simulate('click');
        expect(mockOnValueChange).toHaveBeenCalledTimes(1);
        expect(mockOnValueChange).toHaveBeenCalledWith({ target: { value: {
            type: 'Feature' as const,
            geometry: placeFeature1.geometry,
            properties: { lastAction: 'findPlace', geocodingQueryString: geocodingString }
        }}})

        // There should not be any selection or confirm widgets
        findPlaceWidget.update();
        const selectionList2 = findPlaceWidget.find(`select`);
        expect(selectionList2.length).toEqual(0);
        const confirmButton2 = findPlaceWidget.find({id: `${testId}_confirm`, type: 'button'});
        expect(confirmButton2.length).toEqual(0);
        
    });

    test('Geocode with single result, then re-query with undefined results', async () => {
    
        const findPlaceWidget = mount(<InputMapFindPlace
            id={testId}
            onValueChange={mockOnValueChange}
            widgetConfig={testWidgetConfig}
            value={{ type: 'Feature' as const, properties: {}, geometry: { type: 'Point' as const, coordinates: [-73.1, 45.02]}}}
            inputRef={React.createRef()}
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
            loadingState={0}
        />);

        // Find and click on the Geocode button, to return multiple values
        mockedGeocode.mockResolvedValueOnce([placeFeature1]);
        const geocodeButton = findPlaceWidget.find({id: `${testId}_refresh`, type: 'button'});
        expect(geocodeButton.getDOMNode<HTMLButtonElement>().textContent).toBe('Geocode');
        geocodeButton.simulate('mousedown');
        geocodeButton.simulate('mouseup');
        // Let async functions terminate
        await new Promise(process.nextTick);
    
        // Make sure the widget after 1 result are present
        findPlaceWidget.update();
        expect(mockedGeocode).toHaveBeenCalledTimes(1);
        expect(mockedGeocode).toHaveBeenCalledWith(geocodingString, expect.anything());
        const selectionList = findPlaceWidget.find(`select`);
        expect(selectionList.length).toEqual(1);

        const confirmButton = findPlaceWidget.find({id: `${testId}_confirm`, type: 'button'});
        expect(confirmButton.length).toEqual(1);
        
        // Click the geocode button again, but get undefined values
        mockedGeocode.mockResolvedValueOnce(undefined);
        const newGeocodingString = 'other string';
        testWidgetConfig.geocodingQueryString.mockReturnValueOnce(newGeocodingString);
        geocodeButton.simulate('mousedown');
        geocodeButton.simulate('mouseup');
        // Let async functions terminate
        await new Promise(process.nextTick);
    
        // The select list should display 4 children (2 places and the 2 extra elements) and should not have a confirm button
        findPlaceWidget.update();
        expect(mockedGeocode).toHaveBeenCalledTimes(2);
        expect(mockedGeocode).toHaveBeenLastCalledWith(newGeocodingString, expect.anything());
        const selectionList2 = findPlaceWidget.find(`select`);
        expect(selectionList2.length).toEqual(0);
        const confirmButton2 = findPlaceWidget.find({id: `${testId}_confirm`, type: 'button'});
        expect(confirmButton2.length).toEqual(0);
        
    });

    test('Geocode with single result, then re-query with rejection', async () => {
    
        const findPlaceWidget = mount(<InputMapFindPlace
            id={testId}
            onValueChange={mockOnValueChange}
            widgetConfig={testWidgetConfig}
            value={{ type: 'Feature' as const, properties: {}, geometry: { type: 'Point' as const, coordinates: [-73.1, 45.02]}}}
            inputRef={React.createRef()}
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
            loadingState={0}
        />);

        // Find and click on the Geocode button, to return multiple values
        mockedGeocode.mockResolvedValueOnce([placeFeature1]);
        const geocodeButton = findPlaceWidget.find({id: `${testId}_refresh`, type: 'button'});
        expect(geocodeButton.getDOMNode<HTMLButtonElement>().textContent).toBe('Geocode');
        geocodeButton.simulate('mousedown');
        geocodeButton.simulate('mouseup');
        // Let async functions terminate
        await new Promise(process.nextTick);
    
        // Make sure the widget after 1 result are present
        findPlaceWidget.update();
        expect(mockedGeocode).toHaveBeenCalledTimes(1);
        expect(mockedGeocode).toHaveBeenCalledWith(geocodingString, expect.anything());
        const selectionList = findPlaceWidget.find(`select`);
        expect(selectionList.length).toEqual(1);

        const confirmButton = findPlaceWidget.find({id: `${testId}_confirm`, type: 'button'});
        expect(confirmButton.length).toEqual(1);
        
        // Click the geocode button again, but get undefined values
        mockedGeocode.mockRejectedValueOnce('error geocoding');
        const newGeocodingString = 'other string';
        testWidgetConfig.geocodingQueryString.mockReturnValueOnce(newGeocodingString);
        geocodeButton.simulate('mousedown');
        geocodeButton.simulate('mouseup');
        // Let async functions terminate
        await new Promise(process.nextTick);
    
        // The select list should display 4 children (2 places and the 2 extra elements) and should not have a confirm button
        findPlaceWidget.update();
        expect(mockedGeocode).toHaveBeenCalledTimes(2);
        expect(mockedGeocode).toHaveBeenLastCalledWith(newGeocodingString, expect.anything());
        const selectionList2 = findPlaceWidget.find(`select`);
        expect(selectionList2.length).toEqual(0);
        const confirmButton2 = findPlaceWidget.find({id: `${testId}_confirm`, type: 'button'});
        expect(confirmButton2.length).toEqual(0);
        
    });

    test('Click the geocode button, that triggers an update', async () => {
    
        const props = {
            id: testId,
            onValueChange: mockOnValueChange,
            widgetConfig: testWidgetConfig,
            value: { type: 'Feature' as const, properties: {}, geometry: { type: 'Point' as const, coordinates: [-73.1, 45.02]}},
            inputRef: React.createRef() as React.LegacyRef<HTMLInputElement>,
            size: 'medium' as const,
            interview: interviewAttributes,
            user: userAttributes,
            path: 'foo.test',
            loadingState: 0,
        }

        const findPlaceWidget = mount(<InputMapFindPlace {...props} />);

        // Find and click on the Geocode button, which should trigger an update
        mockedGeocode.mockResolvedValueOnce([placeFeature1]);
        const geocodeButton = findPlaceWidget.find({id: `${testId}_refresh`, type: 'button'});
        expect(geocodeButton.getDOMNode<HTMLButtonElement>().textContent).toBe('Geocode');
        geocodeButton.simulate('mousedown');
        // Update props to simulate update
        findPlaceWidget.setProps({...props, loadingState:1 });
        geocodeButton.simulate('mouseup');
        // Let async functions terminate
        await new Promise(process.nextTick);

        // The geocode function should not have been called
        findPlaceWidget.update();
        expect(mockedGeocode).not.toHaveBeenCalled();

        // Terminate the update, the geocode function should now be called
        findPlaceWidget.setProps({...props, loadingState: 0 });
        // Let async functions terminate
        await new Promise(process.nextTick);

        findPlaceWidget.update();
        expect(mockedGeocode).toHaveBeenCalledTimes(1);
        expect(mockedGeocode).toHaveBeenCalledWith(geocodingString, expect.anything());
        const selectionList = findPlaceWidget.find(`select`);
        expect(selectionList.length).toEqual(1);

        const confirmButton = findPlaceWidget.find({id: `${testId}_confirm`, type: 'button'});
        expect(confirmButton.length).toEqual(1);
        
    });
});
