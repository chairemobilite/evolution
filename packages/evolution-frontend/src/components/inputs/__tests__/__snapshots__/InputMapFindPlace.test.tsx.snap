// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render InputMapPoint with various parameters Test with all parameters 1`] = `
<div
  className="survey-question__input-map-container"
>
  <input
    className="apptr__form-input input-map-hidden"
    defaultValue="{\\"type\\":\\"Feature\\",\\"properties\\":{},\\"geometry\\":{\\"type\\":\\"Point\\",\\"coordinates\\":[-73.1,45.02]}}"
    id="test"
    readOnly={true}
    type="hidden"
  />
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
      }
    }
  >
    <button
      className="button refresh-geocode green large"
      id="test_refresh"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-map-marker-alt fa-w-12 faIconLeft"
        data-icon="map-marker-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M172.268 501.67C26.97 291.031 0 269.413 0 192 0 85.961 85.961 0 192 0s192 85.961 192 192c0 77.413-26.97 99.031-172.268 309.67-9.535 13.774-29.93 13.773-39.464 0zM192 272c44.183 0 80-35.817 80-80s-35.817-80-80-80-80 35.817-80 80 35.817 80 80 80z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      Refresh map
    </button>
    <div
      style={
        Object {
          "marginLeft": 10,
        }
      }
    />
  </div>
  <div
    aria-hidden="true"
    className="survey-question__input-map-container"
  >
    <div
      className="loader"
    >
      <p>
        Loading
        ...
      </p>
    </div>
  </div>
</div>
`;

exports[`Render InputMapPoint with various parameters Test with minimal parameters 1`] = `
<div
  className="survey-question__input-map-container"
>
  <input
    className="apptr__form-input input-map-hidden"
    id="test"
    readOnly={true}
    type="hidden"
  />
  <div
    aria-hidden="true"
    className="survey-question__input-map-container"
  >
    <div
      className="loader"
    >
      <p>
        Loading
        ...
      </p>
    </div>
  </div>
</div>
`;
